### DISCUSSION TITLE
General Networking Questions for AppleMidi

### DISCUSSION THREAD

**Discussion by HonakerM** [lathoub/Arduino-AppleMIDI-Library#111] [Q&A] [answered]
Hello!! I am starting to use this library and I had a few question. I am also new to midi so I'm sorry if my questions aren't applicable.

1. Are clients bidirectional or are outbound clients different then inbound?
1a. From [this documentation](https://github.com/lathoub/Arduino-AppleMIDI-Library/wiki/Memory-footprint#arduino-memory-usage) It says that `MaxNumberOfParticipants` is limited to 4 on `WZ5500` chips. Does this mean 4 inbound connections and 4 outbound?
2. Is there a way to whitelist a set of ip's that are allowed to connect? I'm a bit worried about security and want to restrict what my device can listen to.

Thank you! Any help would be greatly appreciated.

**Comment by lathoub** [comment, marked as answer]
Hello @HonakerM 

1. Yes, clients are bidirectional: the can send and receive, independent of each other. `MaxNumberOfParticipants` is a limit imposed by the W5500 hardware, in practice I recommend not having multiple connections: performance degrades rapidly with multiple connections. 
2. Also yes, check out the [directory](https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/master/examples/AVR_Directory/AVR_Directory.ino) example allowing only connections from addresses registered in the directory (similar behaviour as AppleMIDI on MacOS or Windows)

Have fun learning Arduino and AppleMIDI

**Reply by HonakerM** [reply]
@lathoub Thank you so much for the response! A clarifying question about 1: Can a client only be sender client? E.g. can I send to 2 ips and then receive from 2 different ones?

Another question I had was there a way to force disconnect a client? Thank you again for your help!

**Reply by lathoub** [reply]
> can I send to 2 ips and then receive from 2 different ones?

No, the UDP socket communication is 1-1, MIDI-thru is not supported using AppleMIDI.

> Another question I had was there a way to force disconnect a client? Thank you again for your help!

if you use the Arduino as a server (Session initiator), disconnect via `sendEndSession`
if you use the Arduino as a client (most cases), there is no `end()` method. (relates to [this](https://github.com/lathoub/Arduino-AppleMIDI-Library/discussions/112) recent discussion)

**Reply by HonakerM** [reply]
I'm using it as a server so that's perfect! Thank you
